File: client/src/data/query-configuration.js
import { compileExpression, compileSearch, createRegexBudget } from './filter-expression.js';
import { filterFieldTypes } from './configuration-catalog.js';
import { configurationResource } from './configuration-access.js';
import { ProviderError, canonicalJson } from './data-provider.js';

export function resolveQueryConfiguration(snapshot, input, actor = { id: 'local', capabilities: ['*'] }, options = {}) {
  const definitionVersion = input.definitionVersion === undefined ? 1 : input.definitionVersion;
  if (![1, 2].includes(definitionVersion)) throw new ProviderError('unsupported_query_definition', 'Supported query definition versions are 1 and 2', 422);
  if (definitionVersion === 1 && input.relationshipMode !== undefined) throw new ProviderError('unsupported_query_definition', 'Relationship modes require query definition version 2', 422);
  const filters = input.filters === undefined ? {} : input.filters;
  if (!filters || typeof filters !== 'object' || Array.isArray(filters) || Object.keys(filters).some(key => !['kind', 'sourceId', 'expression', 'sourceIds', 'kinds', 'schemaRefs', 'filterId', 'filterVersion'].includes(key))) throw new ProviderError('invalid_filter', 'Unknown filter field', 422);
  if ((filters.filterId == null) !== (filters.filterVersion == null)) throw new ProviderError('invalid_filter', 'Filter ID and version must occur together', 422);
  let saved;
  if (filters.filterId != null) {
    const resource = configurationResource(snapshot, 'filters', filters.filterId, actor);
    saved = resource.versions.find(item => item.version === filters.filterVersion)?.definition;
    if (!saved) throw new ProviderError('configuration_version_unavailable', 'Saved filter publication is unavailable', 409);
    if (filters.schemaRefs !== undefined && canonicalJson(filters.schemaRefs) !== canonicalJson(saved.schemaRefs)) throw new ProviderError('invalid_filter', 'A saved filter retains its published schema scope', 422);
  }
  if (definitionVersion === 1 && saved?.definitionVersion === 2) throw new ProviderError('unsupported_query_definition', 'Version 2 saved filters require query definition version 2', 422);
  const relationshipMode = input.relationshipMode ?? saved?.relationshipMode ?? 'independent';
  if (!['independent', 'family'].includes(relationshipMode)) throw new ProviderError('invalid_relationship_mode', 'Choose independent records or family context', 422);
  const schemaRefs = saved?.schemaRefs ?? filters.schemaRefs ?? [];
  if (!Array.isArray(schemaRefs) || schemaRefs.length > 100 || schemaRefs.some(pin => !pin || typeof pin.id !== 'string' || !Number.isSafeInteger(pin.version) || pin.version < 1 || Object.keys(pin).some(key => !['id', 'version'].includes(key)))) throw new ProviderError('invalid_filter', 'Schema scope requires bounded exact version pins', 422);
  for (const pin of schemaRefs) configurationResource(snapshot, 'schemas', pin.id, actor);
  const fieldTypes = filterFieldTypes(snapshot, schemaRefs);
  if (definitionVersion === 1 && [filters.expression, saved?.expression].some(expression => expression?.version === 2)) throw new ProviderError('unsupported_query_definition', 'Version 2 expressions require query definition version 2', 422);
  const regexBudget = createRegexBudget(options);
  const predicate = compileExpression(filters.expression, { fieldTypes, regexBudget });
  const savedPredicate = compileExpression(saved?.expression, { fieldTypes, regexBudget });
  const savedSearch = saved?.search;
  const searchInput = saved ? { search: savedSearch.text, searchMode: savedSearch.mode,
    ...(savedSearch.mode === 'regex' ? { searchFlags: savedSearch.flags, searchMatchMode: savedSearch.matchMode, searchDialect: savedSearch.dialect } : { searchCaseSensitive: savedSearch.caseSensitive }),
    searchFields: savedSearch.fields, ...input } : input;
  if (saved && input.searchMode !== undefined && input.searchMode !== savedSearch.mode) {
    for (const key of input.searchMode === 'regex' ? ['searchCaseSensitive'] : ['searchFlags', 'searchMatchMode', 'searchDialect']) {
      if (!(key in input)) delete searchInput[key];
    }
  }
  const search = compileSearch(searchInput, { fieldTypes, regexBudget });
  const required = snapshot.manifest?.legacy?.configuration?.sourcePredicates ?? {};
  if (!required || typeof required !== 'object' || Array.isArray(required) || Object.keys(required).length > 100 || Object.keys(required).some(id => !snapshot.manifest.scope.sourceIds.includes(id))) throw new ProviderError('invalid_source_predicate', 'Source predicates must identify configured sources', 422);
  const sourcePredicates = new Map(Object.entries(required).map(([id, expression]) => [id, compileExpression(expression, { fieldTypes: { ...fieldTypes, '/data/namespace': 'string' }, regexBudget })]));
  const kind = filters.kind === undefined ? 'all' : filters.kind, source = filters.sourceId === undefined ? 'all' : filters.sourceId;
  if (!['all', 'event', 'session'].includes(kind) || typeof source !== 'string') throw new ProviderError('invalid_filter', 'Unknown record kind or source', 422);
  const sourceIds = filters.sourceIds === undefined ? null : filters.sourceIds, kinds = filters.kinds === undefined ? ['event', 'session'] : filters.kinds;
  if (sourceIds !== null && (!Array.isArray(sourceIds) || sourceIds.length > 100 || sourceIds.some(id => typeof id !== 'string') || new Set(sourceIds).size !== sourceIds.length)) throw new ProviderError('invalid_filter', 'Invalid source list', 422);
  if (!Array.isArray(kinds) || kinds.length > 2 || kinds.some(value => !['event', 'session'].includes(value)) || new Set(kinds).size !== kinds.length) throw new ProviderError('invalid_filter', 'Invalid kinds list', 422);
  const schemaKeys = new Set(schemaRefs.map(pin => `${pin.id}:${pin.version}`));
  const sourceSelected = sourceId => (source === 'all' || sourceId === source) && (sourceIds === null || sourceIds.includes(sourceId))
    && (!saved || saved.sourceIds === null || saved.sourceIds.includes(sourceId));
  function hardPredicate(record) {
    return !record.deletedAt && (kind === 'all' || record.kind === kind) && sourceSelected(record.sourceId) && kinds.includes(record.kind)
      && (!saved || saved.kinds.includes(record.kind))
      && (!schemaKeys.size || schemaKeys.has(`${record.schemaId}:${record.schemaVersion}`)) && (!sourcePredicates.has(record.sourceId) || sourcePredicates.get(record.sourceId)(record));
  }
  const directPredicate = record => savedPredicate(record) && predicate(record);
  return { search, fieldTypes, definitionVersion, relationshipMode, sourceSelected, hardPredicate, directPredicate,
    explanationDefinition: { expressions: [saved?.expression, filters.expression].filter(Boolean), search: searchInput },
    predicate: record => hardPredicate(record) && directPredicate(record) };
}

File: server/app/services/query_configuration.py
from ..models.configuration_catalog import _resource_at, filter_field_types
from ..models.domain import DomainError
from .filters import compile_expression, compile_search, create_regex_budget
from .preparation_control import checkpoint


def resolve_query_configuration(snapshot, request, access=None):
    definition_version = request.get('definitionVersion', 1)
    if type(definition_version) is not int or definition_version not in (1, 2):
        raise DomainError('unsupported_query_definition', 'Supported query definition versions are 1 and 2.')
    if definition_version == 1 and 'relationshipMode' in request:
        raise DomainError('unsupported_query_definition', 'Relationship modes require query definition version 2.')
    actor = access["actor"] if access else {"id": "internal", "capabilities": ["*"]}
    filters = request.get("filters", {})
    if not isinstance(filters, dict) or set(filters) - {"kind", "sourceId", "expression", "sourceIds", "kinds", "schemaRefs", "filterId", "filterVersion"}:
        raise DomainError("invalid_filter", "Unknown filter field.")
    if (filters.get("filterId") is None) != (filters.get("filterVersion") is None):
        raise DomainError("invalid_filter", "Filter ID and version must occur together.")
    saved = None
    if filters.get("filterId") is not None:
        resource = _resource_at(snapshot, "filters", filters["filterId"], {"actor": actor})
        saved = next((item["definition"] for item in resource["versions"] if type(filters["filterVersion"]) is int and item["version"] == filters["filterVersion"]), None)
        if saved is None:
            raise DomainError("configuration_version_unavailable", "Saved filter publication is unavailable.", 409)
        if access and saved["sourceIds"] is not None and any(source not in access["sourceIds"] for source in saved["sourceIds"]):
            raise DomainError("configuration_not_found", "Configuration resource is unavailable.", 404)
        if "schemaRefs" in filters and filters["schemaRefs"] != saved["schemaRefs"]:
            raise DomainError("invalid_filter", "A saved filter retains its published schema scope.")
    if definition_version == 1 and saved and saved.get('definitionVersion') == 2:
        raise DomainError('unsupported_query_definition', 'Version 2 saved filters require query definition version 2.')
    relationship_mode = request.get('relationshipMode', saved.get('relationshipMode', 'independent') if saved else 'independent')
    if relationship_mode not in ('independent', 'family'):
        raise DomainError('invalid_relationship_mode', 'Choose independent records or family context.')
    refs = saved["schemaRefs"] if saved is not None else filters.get("schemaRefs", [])
    if (not isinstance(refs, list) or len(refs) > 100 or any(not isinstance(pin, dict) or set(pin) != {"id", "version"}
            or not isinstance(pin["id"], str) or type(pin["version"]) is not int or pin["version"] < 1 for pin in refs)):
        raise DomainError("invalid_filter", "Schema scope requires bounded exact version pins.")
    for pin in refs:
        _resource_at(snapshot, "schemas", pin["id"], {"actor": actor})
    fields = filter_field_types(snapshot, refs)
    if definition_version == 1 and any(isinstance(value, dict) and value.get('version') == 2
                                     for value in (filters.get('expression'), saved.get('expression') if saved else None)):
        raise DomainError('unsupported_query_definition', 'Version 2 expressions require query definition version 2.')
    regex_budget = create_regex_budget(check_cancelled=checkpoint)
    predicate = compile_expression(filters.get("expression"), field_types=fields, regex_budget=regex_budget)
    saved_predicate = compile_expression(saved["expression"] if saved else None, field_types=fields, regex_budget=regex_budget)
    if saved:
        saved_search = saved['search']
        search_input = {'search': saved_search['text'], 'searchMode': saved_search['mode'], 'searchFields': saved_search['fields'],
                        **({'searchFlags': saved_search.get('flags', []), 'searchMatchMode': saved_search.get('matchMode', 'search'),
                            'searchDialect': saved_search.get('dialect', 're2-common-v1')} if saved_search['mode'] == 'regex'
                           else {'searchCaseSensitive': saved_search['caseSensitive']}), **request}
    else:
        search_input = request
    if saved and 'searchMode' in request and request['searchMode'] != saved_search['mode']:
        for key in (('searchCaseSensitive',) if request['searchMode'] == 'regex' else ('searchFlags', 'searchMatchMode', 'searchDialect')):
            if key not in request:
                search_input.pop(key, None)
    search = compile_search(search_input, field_types=fields, regex_budget=regex_budget)
    required = snapshot['manifest'].get('legacy', {}).get('configuration', {}).get('sourcePredicates', {})
    if not isinstance(required, dict) or len(required) > 100 or any(identity not in snapshot['manifest']['scope']['sourceIds'] for identity in required):
        raise DomainError('invalid_source_predicate', 'Source predicates must identify configured sources.')
    source_predicates = {identity: compile_expression(expression, field_types={**fields, '/data/namespace': 'string'}, regex_budget=regex_budget)
                         for identity, expression in required.items()}
    kind, source = filters.get("kind", "all"), filters.get("sourceId", "all")
    if kind not in ("all", "event", "session") or not isinstance(source, str):
        raise DomainError("invalid_filter", "Unknown record kind or source.")
    sources, kinds = filters.get("sourceIds"), filters.get("kinds", ["event", "session"])
    if sources is not None and (not isinstance(sources, list) or len(sources) > 100 or any(not isinstance(item, str) for item in sources) or len(set(sources)) != len(sources)):
        raise DomainError("invalid_filter", "Invalid source list.")
    if not isinstance(kinds, list) or len(kinds) > 2 or any(item not in ("event", "session") for item in kinds) or len(set(kinds)) != len(kinds):
        raise DomainError("invalid_filter", "Invalid kinds list.")
    schema_keys = {(pin["id"], pin["version"]) for pin in refs}

    def source_selected(source_id):
        return ((source == 'all' or source_id == source) and (sources is None or source_id in sources)
                and (saved is None or saved['sourceIds'] is None or source_id in saved['sourceIds']))

    def hard_predicate(record):
        return (record["deletedAt"] is None and (kind == "all" or record["kind"] == kind) and source_selected(record['sourceId']) and record["kind"] in kinds
                and (saved is None or record["kind"] in saved["kinds"])
                and (not schema_keys or (record.get("schemaId"), record.get("schemaVersion")) in schema_keys)
                and (record['sourceId'] not in source_predicates or source_predicates[record['sourceId']](record)))

    def direct_predicate(record):
        return saved_predicate(record) and predicate(record)

    return {"predicate": lambda record: hard_predicate(record) and direct_predicate(record),
            "sourceSelected": source_selected, "hardPredicate": hard_predicate, "directPredicate": direct_predicate,
            "definitionVersion": definition_version, "relationshipMode": relationship_mode,
            'explanationDefinition': {'expressions': [value for value in (saved.get('expression') if saved else None, filters.get('expression')) if value is not None], 'search': search_input},
            "search": search, "fieldTypes": fields}
